src/server/coprocessor/endpoint.rs

```diff
@@ -486,14 +480,163 @@ fn get_chunk(chunks: &mut Vec<Chunk>) -> &mut Chunk {
+            return Err(box_err!("column {} of {} is missing", col_id, h));
+        } else {
+            box_try!(datum::encode_to(data, &[Datum::Null], false));
+        }
+    }
+    Ok(())
+}
+
+#[derive(Debug, Clone)]
+struct SortRow {
+    key: Vec<Datum>,
+    handle: i64,
+    data: Vec<u8>,
+    order_cols: Vec<ByItem>,
+    ctx: Rc<EvalContext>,
+}
+
+
+impl SortRow {
+    fn new(handle: i64,
+           order_cols: Vec<ByItem>,
+           ctx: Rc<EvalContext>,
+           data: Vec<u8>,
+           key: Vec<Datum>)
+           -> SortRow {
+        SortRow {
+            key: key,
+            handle: handle,
+            data: data,
+            order_cols: order_cols,
+            ctx: ctx,
+        }
+    }
+
+    fn cmp_and_check(&self, right: &SortRow) -> Result<CmpOrdering> {
+        let values = self.key.iter().zip(right.key.iter());
+        for (col, (v1, v2)) in self.order_cols.iter().zip(values) {
+            let cmp_ret = v1.cmp(self.ctx.as_ref(), v2);
+            if let StdError::Err(err) = cmp_ret {
+                return Err(box_err!("cmp failed {:?} ", err));
+            }
+
+            let order = cmp_ret.unwrap();
+            if order == CmpOrdering::Equal {
+                continue;
+            }
+            // less or equal
+            if col.get_desc() {
+                // heap pop the biggest first
+                return Ok(order.reverse());
+            } else {
+                return Ok(order);
+            }
+        }
+        Ok(CmpOrdering::Equal)
+    }
+}
+
+struct TopNHeap {
+    rows: BinaryHeap<SortRow>,
+    limit: usize,
+}
+
+impl TopNHeap {
+    fn new(limit: usize) -> TopNHeap {
+        TopNHeap {
+            rows: match limit {
+                usize::MAX => BinaryHeap::new(),
+                _ => BinaryHeap::with_capacity(limit + 1),
+            },
+            limit: limit,
+        }
+    }
+
+    fn try_add_row(&mut self, row: SortRow) -> Result<()> {
+        if self.rows.is_empty() {
+            self.rows.push(row);
+            return Ok(());
+        }
+
+        // check row data
+        let order = {
+            let top_data = self.rows.peek().unwrap();
+            try!(row.cmp_and_check(&top_data))
+        };
+
+        // push into heap on heap not full
+        if self.rows.len() < self.limit {
+            self.rows.push(row);
+            return Ok(());
+        }
+
+        // switch top value with row when heap is full and current row is smaller than top data
+        let mut top_value = self.rows.peek_mut().unwrap();
+        if CmpOrdering::Less == order {
+            *(top_value) = row;
+        }
+        Ok(())
+    }
+
+    fn into_sorted_vec(self) -> Vec<SortRow> {
+        self.rows.into_sorted_vec()
+    }
+}
+
+impl<'a> Ord for SortRow {
+    fn cmp(&self, right: &SortRow) -> CmpOrdering {
+        if let Ok(order) = self.cmp_and_check(right) {
+            return order;
+        }
+        CmpOrdering::Equal
```

Comment:
This step won't be executed since we run `cmp_and_check`  before push data into heap in `try_add_row` @siddontang